src/tlo/analysis/life_expectancy.py

```diff
@@ -0,0 +1,305 @@
+        for i in range((number_age_groups - 2), -1, -1):
+            py_lived_beyond_start_of_interval[i] = py_lived_beyond_start_of_interval[i + 1] + py_lived_in_interval[i]
+
+        # calculate observed life expectancy at start of interval
+        condition = number_alive_at_start_of_interval == 0
+        observed_life_expectancy = pd.Series(index=range(number_age_groups), dtype=float)
+        observed_life_expectancy[condition] = 0
+        observed_life_expectancy[~condition] = py_lived_beyond_start_of_interval / number_alive_at_start_of_interval
+
+        # estimated life expectancy from birth
+        estimated_life_expectancy = estimated_life_expectancy.append(pd.Series([observed_life_expectancy[0]]))
+
+    return estimated_life_expectancy
+
+
+# generate life expectancy estimates for all draws/runs
+def produce_life_expectancy_estimates(results_folder: Path, median=True):
+    """
+    produces sets of life expectancy estimates for each draw/run
+    calls:
+    *1 num_deaths_by_age_group
+    *2 aggregate_person_years_by_age
+
+    Args:
+    - results_folder (PosixPath): The path to the folder containing all logged outputs
+    - median: declare whether to return a median value with 95% uncertainty intervals
+    or return the estimate for each draw/run
+
+    Returns:
+    - pd.DataFrame: The DataFrame with the life expectancy estimates (in years)
+    for every draw/run in the results folder
+    """
+    info = get_scenario_info(results_folder)
+
+    output = pd.DataFrame()
+
+    # extract numbers of deaths
+    deaths = _num_deaths_by_age_group(results_folder)
+
+    # extract person-years
+    person_years = _get_population_size_by_age(results_folder)
+
+    for draw in range(info['number_of_draws']):
+        for run in range(info['runs_per_draw']):
+            # select column
+            le = _estimate_life_expectancy(
+                number_of_deaths_in_interval=deaths.loc[:, (draw, run)],
+                person_years_at_risk=person_years.loc[:, (draw, run)])
+
+            output = pd.concat([output, le], axis=1)
+
+    multi_index_columns = _create_multi_index_columns(results_folder)
+    output.columns = multi_index_columns
+    output.index = ['M', 'F']
+
+    if not median:
+        return output
+
+    else:
+        summary = pd.DataFrame(
+            columns=pd.MultiIndex.from_product(
+                [
+                    output.columns.unique(level='draw'),
+                    ["median", "lower", "upper"]
+                ],
+                names=['draw', 'stat']),
+            index=output.index
+        )
+
+        summary.loc[:, (slice(None), "median")] = output.groupby(axis=1, by='draw').median().values
+        summary.loc[:, (slice(None), "lower")] = output.groupby(axis=1, by='draw').quantile(0.025).values
+        summary.loc[:, (slice(None), "upper")] = output.groupby(axis=1, by='draw').quantile(0.975).values
```

Comment:
wondering if w should use `summarize` for this, as it seems to repeat that logic (?)

Reply:
Yes good point - I've replaced this now